From 962708301d03e9221cf486a4a5b984f15e071bec Mon Sep 17 00:00:00 2001
From: Hien Dang <hien.dang.eb@rvc.renesas.com>
Date: Fri, 18 Mar 2016 20:04:02 +0700
Subject: [PATCH 1293/2066] thermal: rcar_gen3_thermal: Fix reading PTAT from
 THS1

PTAT registers are only available for THS1.
Before this patch, it is wrong assumption that each channel
has its own PTAT registers. This patch will fix this.
Now all channel will read PTAT values from THS1.

Signed-off-by: Hien Dang <hien.dang.eb@rvc.renesas.com>
[zou:Original patch taken from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git
 v4.9/rcar-3.5.1]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 drivers/thermal/rcar_gen3_thermal.c |   18 +++++++++++++++---
 1 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/drivers/thermal/rcar_gen3_thermal.c b/drivers/thermal/rcar_gen3_thermal.c
index bcc2255..18f110a 100644
--- a/drivers/thermal/rcar_gen3_thermal.c
+++ b/drivers/thermal/rcar_gen3_thermal.c
@@ -41,9 +41,12 @@
 #define REG_GEN3_THCODE1	0x50
 #define REG_GEN3_THCODE2	0x54
 #define REG_GEN3_THCODE3	0x58
+
+#define PTAT_BASE		0xE6198000
 #define REG_GEN3_PTAT1		0x5C
 #define REG_GEN3_PTAT2		0x60
 #define REG_GEN3_PTAT3		0x64
+#define PTAT_SIZE		REG_GEN3_PTAT3
 
 /* CTSR bit */
 #define PONM            (0x1 << 8)
@@ -164,12 +167,19 @@ static int round_temp(int temp)
 
 static int thermal_read_fuse_factor(struct rcar_thermal_priv *priv)
 {
+	void __iomem *ptat_base;
 	int err;
 
 	err = RCAR_PRR_INIT();
 	if (err)
 		return err;
 
+	ptat_base = ioremap_nocache(PTAT_BASE, PTAT_SIZE);
+	if (!ptat_base) {
+		dev_err(rcar_priv_to_dev(priv), "Cannot map FUSE register\n");
+		return -ENOMEM;
+	}
+
 	/* For H3 WS1.0, H3 WS1.1 and M3 ES1.0
 	 * these registers have not been programmed yet.
 	 * We will use fixed value as temporary solution.
@@ -207,14 +217,16 @@ static int thermal_read_fuse_factor(struct rcar_thermal_priv *priv)
 		priv->factor.thcode_3 = rcar_thermal_read(priv,
 						REG_GEN3_THCODE3)
 				& GEN3_FUSE_MASK;
-		priv->factor.ptat_1 = rcar_thermal_read(priv, REG_GEN3_PTAT1)
+		priv->factor.ptat_1 = ioread32(ptat_base + REG_GEN3_PTAT1)
 				& GEN3_FUSE_MASK;
-		priv->factor.ptat_2 = rcar_thermal_read(priv, REG_GEN3_PTAT2)
+		priv->factor.ptat_2 = ioread32(ptat_base + REG_GEN3_PTAT2)
 				& GEN3_FUSE_MASK;
-		priv->factor.ptat_3 = rcar_thermal_read(priv, REG_GEN3_PTAT3)
+		priv->factor.ptat_3 = ioread32(ptat_base + REG_GEN3_PTAT3)
 				& GEN3_FUSE_MASK;
 	}
 
+	iounmap(ptat_base);
+
 	return 0;
 }
 
-- 
1.7.5.4

